pkg/pipeline/resource_pipeline_source.go

```diff
@@ -0,0 +1,237 @@
+	BranchExcludePattern string   `json:"branchExcludePattern,omitempty"`
+	BranchIncludePattern string   `json:"branchIncludePattern,omitempty"`
+	Environments         []string `json:"environments,omitempty"`
+	TemplateId           int      `json:"templateId,omitempty"`
+	ID                   int      `json:"id,omitempty"`
+}
+
+const pipelineSourcesUrl = "pipelines/api/v1/pipelinesources"
+
+// func verifyPipelineSource(id string, request *resty.Request) (*resty.Response, error) {
+// 	return request.Head(pipelinesSourcesUrl + id)
+// }
+
+func pipelineSourceResource() *schema.Resource {
+
+	var pipelineSourceSchema = map[string]*schema.Schema{
+		"name": {
+			Type:         schema.TypeString,
+			Required:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
+			Description:  "The name of the pipeline source. Should be prefixed with the project key",
+		},
+		"project_id": {
+			Type:         schema.TypeInt,
+			Required:     true,
+			ValidateFunc: validation.IntAtLeast(0),
+			Description:  "Id of the project where the pipeline source will live.",
+		},
+		"project_integration_id": {
+			Type:         schema.TypeInt,
+			Required:     true,
+			ValidateFunc: validation.IntAtLeast(0),
+			Description:  "Id of the project Github integration to use to create the pipeline source.",
+		},
+		"repository_full_name": {
+			Type:         schema.TypeString,
+			Optional:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
+			Description:  "The full name of the Git repository including the user/organization as it appears in a Git clone command. For example, myOrg/myProject.",
+		},
+		"file_filter": {
+			Type:         schema.TypeString,
+			Required:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
+			Description:  "A regular expression to determine which files to include in pipeline sync (the YML files), with default pipelines.yml. If a templateId was provided, it must be values.yml.",
+		},
+		"is_multi_branch": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Description: "True if the pipeline source is to be a multi-branch pipeline source. Otherwise, it will be a single-branch pipeline source.",
+		},
+		"branch": {
+			Type:         schema.TypeString,
+			Optional:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
+			Description:  "For single branch pipeline sources. Name of branch that has the pipeline definition.",
+		},
+		"branch_exclude_pattern": {
+			Type:         schema.TypeString,
+			Optional:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
+			Description:  "For multi-branch pipeline sources, a regular expression of the branches to exclude.",
+		},
+		"branch_include_pattern": {
+			Type:         schema.TypeString,
+			Optional:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
+			Description:  "For multi-branch pipeline sources, a regular expression of the branches to include.",
+		},
+		"environments": {
+			Type:     schema.TypeList,
```

Comment:
This should be a set. I can't imagine the value of including the same environment multiple times